presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java

```diff
@@ -75,30 +77,41 @@ public static ParquetColumnReader createReader(RichColumnDescriptor descriptor)
             case BOOLEAN:
                 return new ParquetBooleanColumnReader(descriptor);
             case INT32:
-                return new ParquetIntColumnReader(descriptor);
+                return createDecimalColumnReader(descriptor)
+                        .map(Function.identity())
+                        .orElse(new ParquetIntColumnReader(descriptor));
             case INT64:
-                return new ParquetLongColumnReader(descriptor);
+                return createDecimalColumnReader(descriptor)
+                        .map(Function.identity())
+                        .orElse(new ParquetLongColumnReader(descriptor));
             case INT96:
                 return new ParquetTimestampColumnReader(descriptor);
             case FLOAT:
                 return new ParquetFloatColumnReader(descriptor);
             case DOUBLE:
                 return new ParquetDoubleColumnReader(descriptor);
             case BINARY:
-                return new ParquetBinaryColumnReader(descriptor);
+                return createDecimalColumnReader(descriptor)
+                        .map(Function.identity())
+                        .orElse(new ParquetBinaryColumnReader(descriptor));
             case FIXED_LEN_BYTE_ARRAY:
-                if (descriptor.getPrimitiveType().getOriginalType() == OriginalType.DECIMAL) {
-                    DecimalMetadata decimalMetadata = descriptor.getPrimitiveType().getDecimalMetadata();
-                    return ParquetDecimalColumnReaderFactory.createReader(descriptor, decimalMetadata.getPrecision(), decimalMetadata.getScale());
-                }
-                else {
-                    throw new PrestoException(NOT_SUPPORTED, "Parquet type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType());
-                }
+                return createDecimalColumnReader(descriptor)
+                        .map(Function.identity())
+                        .orElseThrow(() -> new PrestoException(NOT_SUPPORTED, "Parquet type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType()));
             default:
                 throw new PrestoException(NOT_SUPPORTED, "Unsupported parquet type: " + descriptor.getType());
         }
     }
 
+    private static Optional<ParquetColumnReader> createDecimalColumnReader(RichColumnDescriptor descriptor)
+    {
+        if (descriptor.getPrimitiveType().getOriginalType() == DECIMAL) {
```

Comment:
I think this will be more readable if you invert the `if` statement and exit early.